flox/core.py

```diff
@@ -1851,17 +1851,27 @@ def groupby_reduce(
     xarray.xarray_reduce
     """
 
+    bys: T_Bys = tuple(np.asarray(b) if not is_duck_array(b) else b for b in by)
+    nby = len(bys)
+    by_is_dask = tuple(is_duck_dask_array(b) for b in bys)
+    any_by_dask = any(by_is_dask)
+
+    if engine is None:
```

Comment:
let's make this a helper function with some nice unit tests:
- sorted `by` -> flox
- unsorted `by` -> numpy
- `by` is dask.
- nD `by`
- `nby` == `
- `argmax` does not give engine="flox"